cmd/promtool/archive.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2018 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"fmt"
+	"os"
+)
+
+const filePerm = 0644
+
+type archiverConfig struct {
+	archiveName string
+}
+
+type archiver interface {
```

Comment:
I'm not sure we need an interface here, since we are not expecting to replace the implementation. Just the struct would do.

But in case you do want to keep the interface, then the constructor method for the concrete type should a) be called after the concrete type, i.e. `tarGzFileWriter` vs. `newArchiver`, b) return the concrete type, not the interface.

Reply:
I was thinking that there might be some situations in the future that they require different ways on writing the information to the files so I want to keep it a `interface` and just change the `newArchiver ` to `newTarGzFileWriter` and it returns `tarGzFileWriter`. Thanks for the suggestion!

Reply:
Fine with me, no biggie. I just wanted to point out that sometimes it's better to not prematurely introduce interfaces before one really needs them.

Reply:
removing the interface might actually remove the naming collisions

Reply:
Hey I know why I did this. I saw [the constructor for `httpClient` in client_golang](https://github.com/prometheus/client_golang/blob/master/api/client.go) and take it as a reference. In this case should it be changed too?

Reply:
Also I did the same thing [here](https://github.com/chyeh/prometheus/blob/PROM-1551/cmd/promtool/http.go#L36) so I'll change it too.

Reply:
yes this is what I mean, but  @chyeh can try different ways and see what works/reads/looks best.